pkg/controller/apicast/apicast_logic_reconciler.go

```diff
@@ -0,0 +1,423 @@
+		Namespace: r.APIcastCR.Namespace,
+	}
+}
+
+func (r *APIcastLogicReconciler) Reconcile() (reconcile.Result, error) {
+	r.Logger().WithValues("Name", r.APIcastCR.Name, "Namespace", r.APIcastCR.Namespace)
+
+	appliedInitialization, err := r.initialize()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if appliedInitialization {
+		return reconcile.Result{}, nil
+	}
+
+	adminPortalCredentialsSecret, changed, err := r.reconcileAdminPortalCredentials()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if changed {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	userProvidedSecrets := &apicastUserProvidedSecrets{
+		adminPortalCredentialsSecret: adminPortalCredentialsSecret,
+	}
+
+	// TODO this function does a little bit of creating the desiredApicast and
+	// also validation. Also validation should be done BEFORE
+	// the initialization probably
+	desiredAPIcast, err := r.internalAPIcast(userProvidedSecrets)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.reconcileDeployment(*desiredAPIcast.Deployment())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.reconcileService(*desiredAPIcast.Service())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if r.APIcastCR.Spec.ExposedHost != nil {
+		err = r.reconcileIngress(*desiredAPIcast.Ingress())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *APIcastLogicReconciler) getAdminPortalCredentialsSecret() (*v1.Secret, error) {
+	adminPortalSecretReference := r.APIcastCR.Spec.AdminPortalCredentialsRef
+	adminPortalNamespace := r.APIcastCR.Namespace
+
+	if adminPortalSecretReference.Namespace != "" {
+		adminPortalNamespace = r.APIcastCR.Spec.AdminPortalCredentialsRef.Namespace
+	}
+
+	if adminPortalSecretReference.Name == "" {
+		return nil, fmt.Errorf("Field 'Name' not specified for AdminPortalCredentialsRef Secret Reference")
+	}
+
+	adminPortalCredentialsNamespacedName := types.NamespacedName{
+		Name:      adminPortalSecretReference.Name,
+		Namespace: adminPortalNamespace,
+	}
+
+	adminPortalCredentialsSecret := v1.Secret{}
+	err := r.Client().Get(context.TODO(), adminPortalCredentialsNamespacedName, &adminPortalCredentialsSecret)
+
+	if err != nil {
+		return nil, err
+	}
+
+	secretStringData := k8sutils.SecretStringDataFromData(adminPortalCredentialsSecret)
+	adminPortalURL, ok := secretStringData[apicast.AdminPortalURLAttributeName]
+	if !ok {
+		return nil, fmt.Errorf("Required key '%s' not found in secret '%s'", apicast.AdminPortalURLAttributeName, adminPortalCredentialsSecret.Name)
+	}
+
+	parsedURL, err := url.Parse(adminPortalURL)
+	if err != nil {
+		return nil, err
+	}
+	_, ok = parsedURL.User.Password()
+	if !ok {
+		return nil, fmt.Errorf("Access Token required in %s URL", apicast.AdminPortalURLAttributeName)
+	}
+
+	return &adminPortalCredentialsSecret, err
+}
+
+func (r *APIcastLogicReconciler) reconcileAdminPortalCredentials() (*v1.Secret, bool, error) {
```

Comment:
This method also returns a bool that is the "changed" flag.
In case OwnerRef is updated we requeue and not continue in this specific case of an "external" user-provided secret